### ISSUE TITLE
Mark TV Show Episodes/Seasons/Series as "Watched"

### ISSUE THREAD

**Issue by TimothyWright95** [jellyfin-archive/jellyfin-roku-legacy#1338]
Ability to mark individual episodes as well as entire seasons/series as watched similar to how you can mark a movie as Watched. This is especially helpful when migrating from plex to jellyfin or when adding other media libraries to jellyfin from another repository.

**Comment by JarodTAerts** [comment]
I agree this would be a good improvement even though it can be done through other clients. It is definitely something I would use. Looking into the code a bit, I feel we can leverage the "options" button on whole seasons to bring up a menu where setting it watched/unwatched is an option. On the induvial episodes the options button is already used for selecting audio and video tracks so we will need to do things a little different. I believe we could add an additional menu to the TVListOptions menu for "Episode Options" or something which includes the options to set watched/unwatched. I think this would make sense from a UX perspective but is adding a very different set of options to that menu which has so far been just audio/video options. @cewert / @1hitsong do you think this path makes sense? If so, I can start working on this improvement.

**Comment by 1hitsong** [comment]
I'd love for us to have the "Mark as Watched" functionality called in the same way regardless of what view the user is on.

Just brainstorming - could we come up with a method to support a long press? The user goes to the item they want, long press on it, then we show a dialog menu that could support options such as Mark As Watched, Add To Playlist, Add To Collection, etc.

**Comment by JarodTAerts** [comment]
Agree consistency will be best. I didn't think of using long press for this, will take a look into how to do that and update on my findings.

**Comment by cewert** [comment]
I'm not against the long press idea but I think the most ideal/consistent solution would be to create a "season page" and an "episode page", then you could have "watched" buttons on the new pages just like with movies. This is how the web client and emby's roku app does it.

Season page could also have a "Play all" button, a favorite button, a shuffle button, and show actors from that season. Episodes could be displayed in a row like on the home screen - play would quick play and OK would open the episode page.

Episode page could have the full description, ratings, audio/subtitle selection, watched button, favorite button, show actors from that episode, and show other episodes from the same season in a row.

**Comment by TimothyWright95** [comment]
I agree with @cewert If you look at how movies already work on the app having a season and episode page makes more sense than a long press. In addition it would be more user friendly as many people probably wouldn't intuitively know that a long press would result in different behavior than a short press; especially on something like a roku app. 

Additionally a season page kindove already exists as when you click a season it brings you to a vertical list of episodes, just adding in some UI updates to change the current season page to include the Watched/Favorite/etc features and perhaps as cewert suggests reorienting the episode list to be horizontal instead of vertical.

**Comment by JarodTAerts** [comment]
Sounds good, I will start taking a look at a broader overhaul of the TV Episodes pages to bring them more in line with the web client and Movies. Will probably take some time but will update this thread when I have something.

**Comment by TimothyWright95** [comment]
Went to watch tv and noticed that an episode detail screen already exists. Hopefully that makes it easier. 

Steps to get to an episode page (only known way that I can tell)
1. Start watching a tv show
2. Exit out halfway through (any point is fine as long as it shows up on your continue watching list)
3. Click the episode on the continue watching list
4. Go down to "Go to episode"

This will get you to the individual episode page in which there is already a mark as watched and other items.

**Comment by cewert** [comment]
@TimothyWright95 Nice catch. That is using the "movie detail screen" and just jamming an episode into it. So yes, it will make it easier in the sense that we have a starting point to copy paste but we still need a separate screen for episodes.

https://github.com/jellyfin/jellyfin-roku/blob/unstable/source/Main.brs#L736

**Comment by ageisen2000** [comment]
I noticed that this works for movies but not TV shows. I'm guessing that it isn't possible to re-use that logic for TV seasons?
